# Real-World Data Cleaning Pipeline

## AgenticX Data Science & Analytics Internship

### Objective
To clean and preprocess a real-world dataset by identifying and handling missing values, duplicate records, inconsistent values, invalid values, incorrect data types, and outliers.

In [68]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [69]:
# Load Dataset
df = pd.read_csv("messy_customer_data.csv")

print("Dataset Loaded Successfully!!")

Dataset Loaded Successfully!!


In [70]:
#Display the first 5 rows of the dataset

df.head(5)

,CustomerID,Name,Gender,Age,City,Signup_Date,Last_purchase_date,purchase_amount,feedback_score,email,Phone_number,Country
0,NaN,Ankit,F,36.0,Mumbai,31-12-2024,13-08-2025,-999.0,2.0,NaN,0,Canada
1,C2,Ravi,Female,NaN,Kolkata,NaN,NaN,NaN,-1.0,user1mail.com,abc123,Canada
2,3,Ravi,female,66.0,Ahmedabad,NaN,20-06-2023,NaN,10.0,user2@mail.com,abc123,India
3,C4,Ankit,male,44.0,Kolkata,NaN,13-09-2025,NaN,NaN,NaN,9316267914,USA
4,5,Rahul,Male,200.0,Ahmedabad,09-07-2025,NaN,-999.0,NaN,user4mail.com,9234603292,USA


In [71]:
#Check the structure and data types of the dataset

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   CustomerID          7621 non-null   str    
 1   Name                8342 non-null   str    
 2   Gender              7545 non-null   str    
 3   Age                 7424 non-null   float64
 4   City                8870 non-null   str    
 5   Signup_Date         7471 non-null   str    
 6   Last_purchase_date  6714 non-null   str    
 7   purchase_amount     4967 non-null   float64
 8   feedback_score      7538 non-null   float64
 9   email               4934 non-null   str    
 10  Phone_number        7556 non-null   str    
 11  Country             8296 non-null   str    
dtypes: float64(3), str(9)
memory usage: 1.4 MB


In [72]:
# Check the number of missing values in each column

missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
CustomerID            2379
Name                  1658
Gender                2455
Age                   2576
City                  1130
Signup_Date           2529
Last_purchase_date    3286
purchase_amount       5033
feedback_score        2462
email                 5066
Phone_number          2444
Country               1704
dtype: int64


In [73]:
#Checking Duplicate values

duplicate_count = df.duplicated().sum()

print("Duplicated_count:",duplicate_count)

Duplicated_count: 0


In [74]:
#Check unique values in categorical columns

categorical_columns = ["Gender","City","Country"]

for column in categorical_columns:
    print("\n---",column,"---")
    print(df[column].value_counts(dropna = False))



--- Gender ---
Gender
NaN       2455
Male      1325
female    1270
F         1262
M         1249
male      1226
Female    1213
Name: count, dtype: int64

--- City ---
City
Hyderabad    1144
Mumbai       1133
Pune         1130
NaN          1130
Bangalore    1120
Chennai      1109
Delhi        1091
Ahmedabad    1087
Kolkata      1056
Name: count, dtype: int64

--- Country ---
Country
NaN          1704
USA          1679
Australia    1660
Canada       1658
UK           1658
India        1641
Name: count, dtype: int64


In [75]:
# Check unique values in numerical columns 

df.describe()

,Age,purchase_amount,feedback_score
count,7424.000000,4967.000000,7538.000000
mean,78.904364,713.654008,4.008092
std,87.817619,2002.187228,4.619336
min,-5.000000,-999.000000,-1.000000
25%,-5.000000,-999.000000,-1.000000
50%,43.000000,-999.000000,3.000000
75%,200.000000,2437.600000,10.000000
max,200.000000,4998.670000,10.000000


In [76]:
# Check the frequency of different age values 

df["Age"].value_counts(dropna = False).sort_index()

Age
-5.0      2506
 18.0       52
 19.0       43
 20.0       45
 21.0       32
 22.0       56
 23.0       42
 24.0       51
 25.0       50
 26.0       43
 27.0       55
 28.0       51
 29.0       51
 30.0       43
 31.0       48
 32.0       44
 33.0       55
 34.0       39
 35.0       47
 36.0       45
 37.0       54
 38.0       67
 39.0       48
 40.0       49
 41.0       48
 42.0       43
 43.0       55
 44.0       57
 45.0       49
 46.0       35
 47.0       51
 48.0       38
 49.0       50
 50.0       45
 51.0       41
 52.0       40
 53.0       36
 54.0       40
 55.0       32
 56.0       41
 57.0       37
 58.0       53
 59.0       60
 60.0       44
 61.0       46
 62.0       39
 63.0       35
 64.0       60
 65.0       43
 66.0       37
 67.0       49
 68.0       49
 69.0       61
 70.0       41
 200.0    2453
 NaN      2576
Name: count, dtype: int64

In [77]:
# Check the frequency of different purchase amount values

df["purchase_amount"].value_counts(dropna = False).head(20)

purchase_amount
 NaN        5033
-999.00     2501
 1478.98       2
 2766.36       2
 1492.24       2
 3381.59       2
 2533.10       2
 3293.82       2
 2051.47       2
 4535.38       2
 4334.90       2
 887.49        1
 2712.50       1
 2822.00       1
 509.03        1
 1582.43       1
 938.55        1
 1250.98       1
 4772.57       1
 84.02         1
Name: count, dtype: int64

In [78]:
# Check the frequency of different feedback score

df["feedback_score"].value_counts(dropna = False).sort_index()

feedback_score
-1.0     2489
 1.0      532
 2.0      495
 3.0      556
 4.0      423
 5.0      522
 10.0    2521
 NaN     2462
Name: count, dtype: int64

In [79]:
df["purchase_amount"] = df["purchase_amount"].replace(-999, np.nan)

In [80]:
# Inspect sample values from the date columns

print("Signup_Date:")
print(df["Signup_Date"].dropna().head(10))

print("\nLast_purchase_date:")
print(df["Last_purchase_date"].dropna().head(10))

Signup_Date:
0     31-12-2024
4     09-07-2025
5     not_a_date
6     not_a_date
8     not_a_date
9     18-12-2023
10    not_a_date
11    29-05-2025
12    not_a_date
13    15-12-2024
Name: Signup_Date, dtype: str

Last_purchase_date:
0     13-08-2025
2     20-06-2023
3     13-09-2025
6        missing
7     15-11-2021
9        missing
11    20-03-2021
13    26-03-2022
14    04-12-2025
16    11-11-2022
Name: Last_purchase_date, dtype: str


In [81]:
# Check CustomerID values and their frequency

print("Sample CustomerID values:")
print(df["CustomerID"].head(20))

print("\nNumber of unique CustomerIDs:")
print(df["CustomerID"].nunique(dropna=True))

print("\nDuplicate CustomerIDs:")
print(df["CustomerID"].duplicated().sum())

Sample CustomerID values:
0     NaN
1      C2
2       3
3      C4
4       5
5      C6
6       7
7       8
8      C9
9      10
10    NaN
11    C12
12     13
13     14
14     15
15    NaN
16     17
17     18
18     19
19     20
Name: CustomerID, dtype: str

Number of unique CustomerIDs:
7621

Duplicate CustomerIDs:
2378


In [82]:
df["CustomerID"].head(20)

0     NaN
1      C2
2       3
3      C4
4       5
5      C6
6       7
7       8
8      C9
9      10
10    NaN
11    C12
12     13
13     14
14     15
15    NaN
16     17
17     18
18     19
19     20
Name: CustomerID, dtype: str

In [83]:
df["CustomerID"].nunique(dropna=True)

7621

In [84]:
df["CustomerID"].duplicated().sum()

np.int64(2378)

In [85]:
# Inspect CustomerID values

print("First 20 CustomerIDs:")
print(df["CustomerID"].head(20).to_list())

print("\nNumber of unique CustomerIDs:")
print(df["CustomerID"].nunique(dropna=True))

print("\nNumber of duplicate CustomerID entries:")
print(df["CustomerID"].duplicated().sum())

print("\nNumber of missing CustomerIDs:")
print(df["CustomerID"].isnull().sum())

First 20 CustomerIDs:
[nan, 'C2', '3', 'C4', '5', 'C6', '7', '8', 'C9', '10', nan, 'C12', '13', '14', '15', nan, '17', '18', '19', '20']

Number of unique CustomerIDs:
7621

Number of duplicate CustomerID entries:
2378

Number of missing CustomerIDs:
2379


In [86]:
# Inspect CustomerIDs that appear more than once

duplicate_ids = df["CustomerID"].value_counts()

print("CustomerIDs appearing more than once:")
print(duplicate_ids[duplicate_ids > 1].head(20))

CustomerIDs appearing more than once:
Series([], Name: count, dtype: int64)


In [87]:
df["CustomerID"].duplicated().sum()

np.int64(2378)

In [88]:
# Check CustomerID format patterns

print("CustomerIDs starting with 'C':")
print(df["CustomerID"].dropna().str.startswith("C").value_counts())

print("\nSample CustomerIDs without 'C':")
print(df.loc[
    df["CustomerID"].notna() &
    ~df["CustomerID"].str.startswith("C"),
    "CustomerID"
].head(20).to_list())

CustomerIDs starting with 'C':
CustomerID
False    5060
True     2561
Name: count, dtype: int64

Sample CustomerIDs without 'C':
['3', '5', '7', '8', '10', '13', '14', '15', '17', '18', '19', '20', '23', '26', '31', '37', '38', '39', '40', '41']


In [89]:
# Inspect email values

print("Sample email values:")
print(df["email"].dropna().head(20).to_list())

print("\nNumber of non-missing emails:")
print(df["email"].notna().sum())

print("\nNumber of unique emails:")
print(df["email"].nunique(dropna=True))

Sample email values:
['user1mail.com', 'user2@mail.com', 'user4mail.com', 'user6mail.com', 'user7@mail.com', 'user9mail.com', 'user10mail.com', 'user12@mail.com', 'user15mail.com', 'user17mail.com', 'user18mail.com', 'user21@mail.com', 'user23mail.com', 'user26mail.com', 'user27mail.com', 'user28@mail.com', 'user30mail.com', 'user32mail.com', 'user36@mail.com', 'user37mail.com']

Number of non-missing emails:
4934

Number of unique emails:
4934


In [90]:
# Check for invalid email formats

email_pattern = r"^[^@\s]+@[^@\s]+\.[^@\s]+$"

invalid_emails = df[
    df["email"].notna() &
    ~df["email"].str.match(email_pattern)
]

print("Number of invalid email values:")
print(len(invalid_emails))

print("\nSample invalid emails:")
print(invalid_emails["email"].head(20).to_list())

Number of invalid email values:
2422

Sample invalid emails:
['user1mail.com', 'user4mail.com', 'user6mail.com', 'user9mail.com', 'user10mail.com', 'user15mail.com', 'user17mail.com', 'user18mail.com', 'user23mail.com', 'user26mail.com', 'user27mail.com', 'user30mail.com', 'user32mail.com', 'user37mail.com', 'user42mail.com', 'user45mail.com', 'user58mail.com', 'user64mail.com', 'user71mail.com', 'user77mail.com']


In [91]:
# Inspect phone number values

print("Sample phone numbers:")
print(df["Phone_number"].dropna().head(20).to_list())

print("\nNumber of non-missing phone numbers:")
print(df["Phone_number"].notna().sum())

print("\nNumber of unique phone numbers:")
print(df["Phone_number"].nunique(dropna=True))

Sample phone numbers:
['0', 'abc123', 'abc123', '9316267914', '9234603292', 'abc123', 'abc123', 'abc123', '0', '0', 'abc123', '0', '9610475090', 'abc123', '0', '9533545457', '9720985722', '9164267579', '0', '0']

Number of non-missing phone numbers:
7556

Number of unique phone numbers:
2512


In [92]:
# Check for invalid phone number formats

phone_pattern = r"^\d{10}$"

invalid_phones = df[
    df["Phone_number"].notna() &
    ~df["Phone_number"].str.match(phone_pattern)
]

print("Number of invalid phone numbers:")
print(len(invalid_phones))

print("\nSample invalid phone numbers:")
print(invalid_phones["Phone_number"].value_counts().head(20))

Number of invalid phone numbers:
5046

Sample invalid phone numbers:
Phone_number
abc123    2586
0         2460
Name: count, dtype: int64


In [93]:
# Check for completely duplicated rows

duplicate_rows = df.duplicated().sum()

print("Number of completely duplicated rows:")
print(duplicate_rows)

Number of completely duplicated rows:
0


## Data Quality Issues Identified

During the initial data quality audit, the following issues were identified:

1. **Missing values**
   - Several columns contain missing values, including CustomerID, Name, Gender, Age, City, Signup_Date, Last_purchase_date, purchase_amount, feedback_score, email, Phone_number, and Country.

2. **Inconsistent Gender values**
   - Gender is represented using multiple formats such as `M`, `Male`, `male`, `F`, `Female`, and `female`.

3. **Invalid Age values**
   - The Age column contains invalid values such as `-5` and `200`, which fall outside a reasonable human-age range.

4. **Invalid Feedback Scores**
   - The feedback_score column contains values such as `-1` and `10`.
   - The observed valid rating range is 1–5.

5. **Placeholder values in purchase_amount**
   - The value `-999` occurs frequently and is treated as a placeholder/missing value rather than a genuine purchase amount.

6. **Invalid date values**
   - Signup_Date contains values such as `not_a_date`.
   - Last_purchase_date contains values such as `missing`.
   - The date columns are currently stored as strings.

7. **Inconsistent CustomerID formatting**
   - CustomerIDs appear in mixed formats, such as `C2`, `C4`, `3`, and `5`.
   - No duplicate non-missing CustomerIDs were identified.

8. **Invalid email formats**
   - Some email values are missing the `@` symbol or otherwise do not match the expected email structure.

9. **Invalid phone numbers**
   - Some Phone_number values contain invalid entries such as `abc123` and `0`.
   - Valid-looking entries contain 10 digits.

10. **Duplicate records**
    - No completely duplicated rows were found in the dataset.

In [94]:
# Standardize Gender values

df["Gender"] = df["Gender"].replace({
    "M": "Male",
    "male": "Male",
    "F": "Female",
    "female": "Female"
})

print("Gender values after standardization:")
print(df["Gender"].value_counts(dropna=False))

Gender values after standardization:
Gender
Male      3800
Female    3745
NaN       2455
Name: count, dtype: int64


In [95]:
# Replace invalid age values with missing values

df.loc[(df["Age"] < 0) | (df["Age"] > 100), "Age"] = np.nan

print("Invalid age values replaced with NaN.")
print("\nAge summary after cleaning:")
print(df["Age"].describe())

Invalid age values replaced with NaN.

Age summary after cleaning:
count    2465.000000
mean       43.698174
std        15.304239
min        18.000000
25%        31.000000
50%        43.000000
75%        58.000000
max        70.000000
Name: Age, dtype: float64


In [96]:
# Replace invalid feedback scores with missing values

df.loc[
    (df["feedback_score"] < 1) | (df["feedback_score"] > 5),
    "feedback_score"
] = np.nan

print("Invalid feedback scores replaced with NaN.")

print("\nFeedback score values after cleaning:")
print(df["feedback_score"].value_counts(dropna=False).sort_index())

Invalid feedback scores replaced with NaN.

Feedback score values after cleaning:
feedback_score
1.0     532
2.0     495
3.0     556
4.0     423
5.0     522
NaN    7472
Name: count, dtype: int64


In [97]:
# Replace the placeholder purchase amount with a missing value

df["purchase_amount"] = df["purchase_amount"].replace(-999, np.nan)

print("Placeholder purchase amounts replaced with NaN.")

print("\nPurchase amount summary after cleaning:")
print(df["purchase_amount"].describe())

Placeholder purchase amounts replaced with NaN.

Purchase amount summary after cleaning:
count    2466.000000
mean     2450.615758
std      1442.834603
min        10.310000
25%      1159.385000
50%      2460.355000
75%      3733.880000
max      4998.670000
Name: purchase_amount, dtype: float64


In [98]:
# Convert Signup_Date to datetime

df["Signup_Date"] = pd.to_datetime(
    df["Signup_Date"],
    format="%d-%m-%Y",
    errors="coerce"
)

print("Signup_Date data type:")
print(df["Signup_Date"].dtype)

print("\nSample Signup_Date values:")
print(df["Signup_Date"].head(10))

Signup_Date data type:
datetime64[us]

Sample Signup_Date values:
0   2024-12-31
1          NaT
2          NaT
3          NaT
4   2025-07-09
5          NaT
6          NaT
7          NaT
8          NaT
9   2023-12-18
Name: Signup_Date, dtype: datetime64[us]


In [99]:
# Convert Last_purchase_date to datetime

df["Last_purchase_date"] = pd.to_datetime(
    df["Last_purchase_date"],
    format="%d-%m-%Y",
    errors="coerce"
)

print("Last_purchase_date data type:")
print(df["Last_purchase_date"].dtype)

print("\nSample Last_purchase_date values:")
print(df["Last_purchase_date"].head(10))

Last_purchase_date data type:
datetime64[us]

Sample Last_purchase_date values:
0   2025-08-13
1          NaT
2   2023-06-20
3   2025-09-13
4          NaT
5          NaT
6          NaT
7   2021-11-15
8          NaT
9          NaT
Name: Last_purchase_date, dtype: datetime64[us]


In [100]:
# Identify and replace invalid email addresses

email_pattern = r"^[^@\s]+@[^@\s]+\.[^@\s]+$"

invalid_email_mask = (
    df["email"].notna() &
    ~df["email"].str.match(email_pattern)
)

df.loc[invalid_email_mask, "email"] = np.nan

print("Invalid email addresses replaced with NaN.")

print("\nNumber of missing emails after cleaning:")
print(df["email"].isna().sum())

Invalid email addresses replaced with NaN.

Number of missing emails after cleaning:
7488


In [101]:
# Identify and replace invalid phone numbers

phone_pattern = r"^\d{10}$"

invalid_phone_mask = (
    df["Phone_number"].notna() &
    ~df["Phone_number"].str.match(phone_pattern)
)

df.loc[invalid_phone_mask, "Phone_number"] = np.nan

print("Invalid phone numbers replaced with NaN.")

print("\nNumber of missing phone numbers after cleaning:")
print(df["Phone_number"].isna().sum())

Invalid phone numbers replaced with NaN.

Number of missing phone numbers after cleaning:
7490


In [102]:
# Check missing values after cleaning invalid data

missing_after_cleaning = df.isnull().sum()

print("Missing values after cleaning:")
print(missing_after_cleaning)

Missing values after cleaning:
CustomerID            2379
Name                  1658
Gender                2455
Age                   7535
City                  1130
Signup_Date           5004
Last_purchase_date    6652
purchase_amount       7534
feedback_score        7472
email                 7488
Phone_number          7490
Country               1704
dtype: int64


In [103]:
# Handle missing categorical values

categorical_missing_columns = ["Gender", "City", "Country"]

for column in categorical_missing_columns:
    df[column] = df[column].fillna("Unknown")

print("Missing categorical values handled.")

print("\nMissing values after categorical cleaning:")
print(df[categorical_missing_columns].isnull().sum())

Missing categorical values handled.

Missing values after categorical cleaning:
Gender     0
City       0
Country    0
dtype: int64


In [104]:
# Calculate median values for numeric columns

age_median = df["Age"].median()
purchase_median = df["purchase_amount"].median()
feedback_median = df["feedback_score"].median()

print("Age median:", age_median)
print("Purchase amount median:", purchase_median)
print("Feedback score median:", feedback_median)

Age median: 43.0
Purchase amount median: 2460.355
Feedback score median: 3.0


In [105]:
# Fill missing numeric values using median

df["Age"] = df["Age"].fillna(age_median)
df["purchase_amount"] = df["purchase_amount"].fillna(purchase_median)
df["feedback_score"] = df["feedback_score"].fillna(feedback_median)

print("Missing numeric values filled using median.")

print("\nMissing values after numeric imputation:")
print(df[["Age", "purchase_amount", "feedback_score"]].isnull().sum())

Missing numeric values filled using median.

Missing values after numeric imputation:
Age                0
purchase_amount    0
feedback_score     0
dtype: int64


In [106]:
# Check remaining missing values

remaining_missing = df.isnull().sum()

print("Remaining missing values:")
print(remaining_missing)

Remaining missing values:
CustomerID            2379
Name                  1658
Gender                   0
Age                      0
City                     0
Signup_Date           5004
Last_purchase_date    6652
purchase_amount          0
feedback_score           0
email                 7488
Phone_number          7490
Country                  0
dtype: int64


In [107]:
# Handle missing customer names

df["Name"] = df["Name"].fillna("Unknown")

print("Missing customer names handled.")

print("\nMissing values in Name:")
print(df["Name"].isnull().sum())

Missing customer names handled.

Missing values in Name:
0


In [108]:
# Check for completely duplicated rows after cleaning

duplicate_rows = df.duplicated().sum()

print("Number of completely duplicated rows:")
print(duplicate_rows)

Number of completely duplicated rows:
17


In [109]:
# Inspect completely duplicated rows

duplicate_mask = df.duplicated(keep=False)

print("Total rows involved in duplicates:")
print(duplicate_mask.sum())

print("\nDuplicate rows:")
display(df[duplicate_mask].sort_values(
    by=["Name", "Age", "City"]
))

Total rows involved in duplicates:
34

Duplicate rows:


,CustomerID,Name,Gender,Age,City,Signup_Date,Last_purchase_date,purchase_amount,feedback_score,email,Phone_number,Country
1474,NaN,Amit,Female,43.0,Ahmedabad,NaT,NaT,2460.355,3.0,NaN,NaN,Canada
1597,NaN,Amit,Female,43.0,Ahmedabad,NaT,NaT,2460.355,3.0,NaN,NaN,Canada
1573,NaN,Amit,Female,43.0,Bangalore,NaT,NaT,2460.355,3.0,NaN,NaN,India
5384,NaN,Amit,Female,43.0,Bangalore,NaT,NaT,2460.355,3.0,NaN,NaN,India
627,NaN,Ankit,Female,43.0,Ahmedabad,NaT,NaT,2460.355,3.0,NaN,NaN,Australia
5465,NaN,Ankit,Female,43.0,Ahmedabad,NaT,NaT,2460.355,3.0,NaN,NaN,Australia
122,NaN,Ankit,Unknown,43.0,Chennai,NaT,NaT,2460.355,3.0,NaN,NaN,USA
197,NaN,Ankit,Unknown,43.0,Chennai,NaT,NaT,2460.355,3.0,NaN,NaN,USA
3097,NaN,Ankit,Male,43.0,Mumbai,NaT,NaT,2460.355,3.0,NaN,NaN,Canada
9925,NaN,Ankit,Male,43.0,Mumbai,NaT,NaT,2460.355,3.0,NaN,NaN,Canada


In [110]:
# Remove completely duplicated rows

before_duplicates = len(df)

df = df.drop_duplicates()

after_duplicates = len(df)

print("Rows before removing duplicates:", before_duplicates)
print("Rows after removing duplicates:", after_duplicates)
print("Duplicate rows removed:", before_duplicates - after_duplicates)

Rows before removing duplicates: 10000
Rows after removing duplicates: 9983
Duplicate rows removed: 17


In [111]:
# Verify that no completely duplicated rows remain

print("Remaining duplicate rows:")
print(df.duplicated().sum())

Remaining duplicate rows:
0


In [112]:
# Detect potential outliers in purchase_amount using IQR

Q1 = df["purchase_amount"].quantile(0.25)
Q3 = df["purchase_amount"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = df[
    (df["purchase_amount"] < lower_bound) |
    (df["purchase_amount"] > upper_bound)
]

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower bound:", lower_bound)
print("Upper bound:", upper_bound)
print("Number of potential outliers:", len(outliers))

Q1: 2460.355
Q3: 2460.355
IQR: 0.0
Lower bound: 2460.355
Upper bound: 2460.355
Number of potential outliers: 2466


In [113]:
df["purchase_amount"] = df["purchase_amount"].fillna(2460.355)

In [114]:
# Inspect the distribution of purchase_amount after cleaning

print("Purchase amount statistics:")
print(df["purchase_amount"].describe())

print("\nNumber of values equal to the median:")
print((df["purchase_amount"] == purchase_median).sum())

Purchase amount statistics:
count    9983.000000
mean     2457.949213
std       717.007472
min        10.310000
25%      2460.355000
50%      2460.355000
75%      2460.355000
max      4998.670000
Name: purchase_amount, dtype: float64

Number of values equal to the median:
7517


In [115]:
# Select original purchase amounts for outlier analysis
# Exclude values that were created through median imputation

original_purchase = df.loc[
    df["purchase_amount"] != purchase_median,
    "purchase_amount"
]

print("Number of original purchase values:")
print(len(original_purchase))

print("\nOriginal purchase amount statistics:")
print(original_purchase.describe())

Number of original purchase values:
2466

Original purchase amount statistics:
count    2466.000000
mean     2450.615758
std      1442.834603
min        10.310000
25%      1159.385000
50%      2460.355000
75%      3733.880000
max      4998.670000
Name: purchase_amount, dtype: float64


In [116]:
# Calculate IQR boundaries using original purchase values

Q1_purchase = original_purchase.quantile(0.25)
Q3_purchase = original_purchase.quantile(0.75)

IQR_purchase = Q3_purchase - Q1_purchase

lower_bound_purchase = Q1_purchase - 1.5 * IQR_purchase
upper_bound_purchase = Q3_purchase + 1.5 * IQR_purchase

print("Q1:", Q1_purchase)
print("Q3:", Q3_purchase)
print("IQR:", IQR_purchase)
print("Lower bound:", lower_bound_purchase)
print("Upper bound:", upper_bound_purchase)

potential_outliers = original_purchase[
    (original_purchase < lower_bound_purchase) |
    (original_purchase > upper_bound_purchase)
]

print("\nNumber of potential outliers:", len(potential_outliers))

Q1: 1159.385
Q3: 3733.88
IQR: 2574.495
Lower bound: -2702.3575
Upper bound: 7595.6224999999995

Number of potential outliers: 0


## Outlier Analysis

Outlier detection was performed on the original observed purchase amounts before median imputation.

The Interquartile Range (IQR) method was used:

- Q1 = 1159.385
- Q3 = 3733.880
- IQR = 2574.495
- Lower bound = -2702.3575
- Upper bound = 7595.6225

The observed purchase amounts range from 10.31 to 4998.67. Since no value falls outside the calculated IQR boundaries, no potential outliers were detected.

Therefore, no purchase amount values were removed as outliers.

In [117]:
# Load the original dataset again for Age outlier analysis

raw_df = pd.read_csv("messy_customer_data.csv")

original_age = raw_df["Age"].dropna()

print("Number of original Age values:", len(original_age))
print("\nOriginal Age statistics:")
print(original_age.describe())

Number of original Age values: 7424

Original Age statistics:
count    7424.000000
mean       78.904364
std        87.817619
min        -5.000000
25%        -5.000000
50%        43.000000
75%       200.000000
max       200.000000
Name: Age, dtype: float64


In [118]:
# Keep only valid human ages for outlier analysis

valid_age = original_age[
    (original_age >= 18) &
    (original_age <= 100)
]

print("Number of valid age values:", len(valid_age))

print("\nValid age statistics:")
print(valid_age.describe())

Number of valid age values: 2465

Valid age statistics:
count    2465.000000
mean       43.698174
std        15.304239
min        18.000000
25%        31.000000
50%        43.000000
75%        58.000000
max        70.000000
Name: Age, dtype: float64


In [119]:
# Calculate IQR boundaries for valid ages

Q1_age = valid_age.quantile(0.25)
Q3_age = valid_age.quantile(0.75)

IQR_age = Q3_age - Q1_age

lower_bound_age = Q1_age - 1.5 * IQR_age
upper_bound_age = Q3_age + 1.5 * IQR_age

print("Q1:", Q1_age)
print("Q3:", Q3_age)
print("IQR:", IQR_age)
print("Lower bound:", lower_bound_age)
print("Upper bound:", upper_bound_age)

age_outliers = valid_age[
    (valid_age < lower_bound_age) |
    (valid_age > upper_bound_age)
]

print("\nNumber of potential age outliers:", len(age_outliers))

Q1: 31.0
Q3: 58.0
IQR: 27.0
Lower bound: -9.5
Upper bound: 98.5

Number of potential age outliers: 0


### Age Outlier Analysis

For outlier analysis, only valid ages between 18 and 100 were considered. The clearly invalid values `-5` and `200` were excluded because they had already been identified as invalid data values.

Using the IQR method:

- Q1 = 31
- Q3 = 58
- IQR = 27
- Lower bound = -9.5
- Upper bound = 98.5

The valid ages range from 18 to 70, so no valid age falls outside the calculated IQR boundaries.

Therefore, no additional statistical outliers were detected in the valid Age values.

In [120]:
# Load original feedback scores for outlier analysis

original_feedback = raw_df["feedback_score"].dropna()

print("Number of original feedback scores:", len(original_feedback))

print("\nOriginal feedback statistics:")
print(original_feedback.describe())

Number of original feedback scores: 7538

Original feedback statistics:
count    7538.000000
mean        4.008092
std         4.619336
min        -1.000000
25%        -1.000000
50%         3.000000
75%        10.000000
max        10.000000
Name: feedback_score, dtype: float64


In [121]:
# Keep only valid feedback scores for outlier analysis

valid_feedback = original_feedback[
    (original_feedback >= 1) &
    (original_feedback <= 5)
]

print("Number of valid feedback scores:", len(valid_feedback))

print("\nValid feedback statistics:")
print(valid_feedback.describe())

Number of valid feedback scores: 2528

Valid feedback statistics:
count    2528.000000
mean        2.963608
std         1.424898
min         1.000000
25%         2.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: feedback_score, dtype: float64


In [122]:
# Calculate IQR boundaries for valid feedback scores

Q1_feedback = valid_feedback.quantile(0.25)
Q3_feedback = valid_feedback.quantile(0.75)

IQR_feedback = Q3_feedback - Q1_feedback

lower_bound_feedback = Q1_feedback - 1.5 * IQR_feedback
upper_bound_feedback = Q3_feedback + 1.5 * IQR_feedback

print("Q1:", Q1_feedback)
print("Q3:", Q3_feedback)
print("IQR:", IQR_feedback)
print("Lower bound:", lower_bound_feedback)
print("Upper bound:", upper_bound_feedback)

feedback_outliers = valid_feedback[
    (valid_feedback < lower_bound_feedback) |
    (valid_feedback > upper_bound_feedback)
]

print("\nNumber of potential feedback outliers:", len(feedback_outliers))

Q1: 2.0
Q3: 4.0
IQR: 2.0
Lower bound: -1.0
Upper bound: 7.0

Number of potential feedback outliers: 0


In [123]:
# Final dataset overview

print("Final dataset shape:")
print(df.shape)

print("\nFinal data types:")
print(df.dtypes)

print("\nRemaining missing values:")
print(df.isnull().sum())

print("\nRemaining duplicate rows:")
print(df.duplicated().sum())

Final dataset shape:
(9983, 12)

Final data types:
CustomerID                       str
Name                             str
Gender                           str
Age                          float64
City                             str
Signup_Date           datetime64[us]
Last_purchase_date    datetime64[us]
purchase_amount              float64
feedback_score               float64
email                            str
Phone_number                     str
Country                          str
dtype: object

Remaining missing values:
CustomerID            2362
Name                     0
Gender                   0
Age                      0
City                     0
Signup_Date           4987
Last_purchase_date    6635
purchase_amount          0
feedback_score           0
email                 7471
Phone_number          7473
Country                  0
dtype: int64

Remaining duplicate rows:
0


In [124]:
# Verify cleaned values and ranges

print("Age range:")
print(df["Age"].min(), "to", df["Age"].max())

print("\nPurchase amount range:")
print(df["purchase_amount"].min(), "to", df["purchase_amount"].max())

print("\nFeedback score range:")
print(df["feedback_score"].min(), "to", df["feedback_score"].max())

print("\nGender values:")
print(df["Gender"].value_counts())

print("\nInvalid purchase placeholders remaining:")
print((df["purchase_amount"] == -999).sum())

print("\nInvalid feedback values remaining:")
print(((df["feedback_score"] < 1) | (df["feedback_score"] > 5)).sum())

Age range:
18.0 to 70.0

Purchase amount range:
10.31 to 4998.67

Feedback score range:
1.0 to 5.0

Gender values:
Gender
Male       3796
Female     3735
Unknown    2452
Name: count, dtype: int64

Invalid purchase placeholders remaining:
0

Invalid feedback values remaining:
0


In [125]:
# Final before-and-after dataset summary

print("BEFORE CLEANING")
print("----------------")
print("Rows:", len(raw_df))
print("Columns:", len(raw_df.columns))
print("Duplicate rows:", raw_df.duplicated().sum())
print("Missing values:", raw_df.isnull().sum().sum())

print("\nAFTER CLEANING")
print("---------------")
print("Rows:", len(df))
print("Columns:", len(df.columns))
print("Duplicate rows:", df.duplicated().sum())
print("Missing values:", df.isnull().sum().sum())

BEFORE CLEANING
----------------
Rows: 10000
Columns: 12
Duplicate rows: 0
Missing values: 32722

AFTER CLEANING
---------------
Rows: 9983
Columns: 12
Duplicate rows: 0
Missing values: 28928


## Before and After Cleaning Summary

The dataset originally contained 10,000 rows and 12 columns with 32,722 missing values.

After applying the data cleaning and preprocessing steps:

- The dataset contains 9,983 rows and 12 columns.
- 17 completely duplicated rows were removed.
- Missing categorical values were handled using `Unknown`.
- Missing numeric values were handled using median imputation.
- Invalid age, feedback, email, phone, and purchase placeholder values were handled appropriately.
- Date columns were converted from strings to datetime format.
- No duplicate rows remain after cleaning.
- Remaining missing values were retained where the original information could not be reliably inferred or reconstructed.

The final dataset was validated for data types, value ranges, invalid values, missing values, and duplicate records.

In [126]:
# Save the cleaned dataset

output_file = "cleaned_customer_data.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully!")
print("File name:", output_file)

Cleaned dataset saved successfully!
File name: cleaned_customer_data.csv


In [127]:
# Reload the cleaned CSV to verify it was saved correctly

cleaned_df = pd.read_csv("cleaned_customer_data.csv")

print("Cleaned CSV reloaded successfully!")

print("\nShape:")
print(cleaned_df.shape)

print("\nColumns:")
print(cleaned_df.columns.tolist())

print("\nFirst 5 rows:")
display(cleaned_df.head())

Cleaned CSV reloaded successfully!

Shape:
(9983, 12)

Columns:
['CustomerID', 'Name', 'Gender', 'Age', 'City', 'Signup_Date', 'Last_purchase_date', 'purchase_amount', 'feedback_score', 'email', 'Phone_number', 'Country']

First 5 rows:


,CustomerID,Name,Gender,Age,City,Signup_Date,Last_purchase_date,purchase_amount,feedback_score,email,Phone_number,Country
0,NaN,Ankit,Female,36.0,Mumbai,2024-12-31,2025-08-13,2460.355,2.0,NaN,NaN,Canada
1,C2,Ravi,Female,43.0,Kolkata,NaN,NaN,2460.355,3.0,NaN,NaN,Canada
2,3,Ravi,Female,66.0,Ahmedabad,NaN,2023-06-20,2460.355,3.0,user2@mail.com,NaN,India
3,C4,Ankit,Male,44.0,Kolkata,NaN,2025-09-13,2460.355,3.0,NaN,9.316268e+09,USA
4,5,Rahul,Male,43.0,Ahmedabad,2025-07-09,NaN,2460.355,3.0,NaN,9.234603e+09,USA


In [128]:
# Keep phone numbers as text when reloading the cleaned CSV

cleaned_df["Phone_number"] = cleaned_df["Phone_number"].astype("string")

print("Phone_number data type:")
print(cleaned_df["Phone_number"].dtype)

print("\nSample phone numbers:")
print(cleaned_df["Phone_number"].dropna().head(10).to_list())

Phone_number data type:
string

Sample phone numbers:
['9316267914.0', '9234603292.0', '9610475090.0', '9533545457.0', '9720985722.0', '9164267579.0', '9693394106.0', '9919454989.0', '9281075045.0', '9168733414.0']


In [129]:
# Remove the unwanted .0 from phone numbers

cleaned_df["Phone_number"] = (
    cleaned_df["Phone_number"]
    .str.replace(r"\.0$", "", regex=True)
)

print("Phone numbers after formatting:")
print(cleaned_df["Phone_number"].dropna().head(10).to_list())

Phone numbers after formatting:
['9316267914', '9234603292', '9610475090', '9533545457', '9720985722', '9164267579', '9693394106', '9919454989', '9281075045', '9168733414']


In [130]:
# Final validation of the cleaned dataset

print("FINAL VALIDATION")
print("================")

print("\nShape:")
print(cleaned_df.shape)

print("\nMissing values:")
print(cleaned_df.isnull().sum())

print("\nDuplicate rows:")
print(cleaned_df.duplicated().sum())

print("\nAge range:")
print(cleaned_df["Age"].min(), "to", cleaned_df["Age"].max())

print("\nPurchase amount range:")
print(cleaned_df["purchase_amount"].min(), "to", cleaned_df["purchase_amount"].max())

print("\nFeedback score range:")
print(cleaned_df["feedback_score"].min(), "to", cleaned_df["feedback_score"].max())

print("\nInvalid phone numbers:")
phone_check = (
    cleaned_df["Phone_number"].notna() &
    ~cleaned_df["Phone_number"].str.match(r"^\d{10}$")
)
print(phone_check.sum())

print("\nInvalid email addresses:")
email_check = (
    cleaned_df["email"].notna() &
    ~cleaned_df["email"].str.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
)
print(email_check.sum())

FINAL VALIDATION

Shape:
(9983, 12)

Missing values:
CustomerID            2362
Name                     0
Gender                   0
Age                      0
City                     0
Signup_Date           4987
Last_purchase_date    6635
purchase_amount          0
feedback_score           0
email                 7471
Phone_number          7473
Country                  0
dtype: int64

Duplicate rows:
0

Age range:
18.0 to 70.0

Purchase amount range:
10.31 to 4998.67

Feedback score range:
1.0 to 5.0

Invalid phone numbers:
0

Invalid email addresses:
0


In [131]:
# Save the final validated cleaned dataset

final_output = "cleaned_customer_data_final.csv"

cleaned_df.to_csv(final_output, index=False)

print("Final cleaned dataset saved successfully!")
print("File name:", final_output)
print("Rows:", len(cleaned_df))
print("Columns:", len(cleaned_df.columns))

Final cleaned dataset saved successfully!
File name: cleaned_customer_data_final.csv
Rows: 9983
Columns: 12


# Conclusion

The messy customer dataset was successfully cleaned and validated through a structured data preprocessing pipeline.

### Key cleaning steps performed:

- Identified and analyzed missing values.
- Standardized inconsistent Gender values.
- Replaced invalid Age values with missing values and applied median imputation.
- Replaced invalid feedback scores outside the 1–5 range with missing values and applied median imputation.
- Replaced the `-999` purchase amount placeholder with missing values and applied median imputation.
- Converted Signup_Date and Last_purchase_date from strings to datetime format.
- Identified and handled invalid email addresses.
- Identified and handled invalid phone numbers.
- Preserved missing CustomerIDs instead of creating artificial identifiers.
- Handled missing categorical and customer-name values appropriately.
- Identified and removed 17 completely duplicated records.
- Performed IQR-based outlier analysis on valid numeric observations.
- Validated the cleaned dataset for ranges, data types, invalid values, missing values, and duplicate records.
- Exported the final cleaned dataset as `cleaned_customer_data_final.csv`.

### Final Dataset

- **Rows:** 9,983
- **Columns:** 12
- **Duplicate rows:** 0
- **Valid Age range:** 18–70
- **Valid feedback score range:** 1–5
- **Invalid phone numbers:** 0
- **Invalid email addresses:** 0

The final dataset is ready for further exploratory data analysis and machine learning workflows.